src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncBatchingWorkQueue.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A queue where items can be added to to be processed in batches after some delay has passed.
+    /// When processing happens, all the items added since the last processing point will be passed
+    /// along to be worked on.  Rounds of processing happen serially, only starting up after a
+    /// previous round has completed.
+    /// </summary>
+    internal class AsyncBatchingWorkQueue<TItem>
```

Comment:
Extracted from https://github.com/dotnet/roslyn/pull/42346

Reply:
Hmm I am unable find this file or functionality in that PR - is this shared utility between designer attribute incremental analyzer and FAR now?

Reply:
in that PR it's embedded in a type.  i'm pulling it out so it can be shared with multiple services.  it's also pulled out in https://github.com/dotnet/roslyn/pull/42383